src/resource.h

```diff
@@ -79,6 +81,16 @@ class Resource {
   /// @return a new resource object with same state id and quantity == quantity
   virtual Ptr ExtractRes(double quantity) = 0;
 
+  /// To enable the Decay method to be called on any child resource, define
+  /// a null op Decay method here.
+  /// @param curr_time the current time for the decay oepration
+  virtual void Decay(int curr_time) { throw Error("cannot decay resource type " + this->type()); };
```

Comment:
In `material.h` you applied a default value of -1. Did you mean to do that in this file too?

Reply:
@gonuke this is from #1687

Reply:
It doesn't matter here, because if the virtual function is not overwritten by a child class, it will invoke this version and throw an exception.